RFC: Add documentation page about error handling and logging. (#2681)
neondatabase/neon: docs/error-handling.md

# Error handling and logging

## Logging errors

The principle is that errors are logged when they are handled. If you
just propagate an error to the caller in a function, you don't need to

> log it; the caller will. But if you "swallow" an error in a function,

Reviewer: "consume" might be a better term than "swallow"

> you *must* log it (if it needs to be logged at all).

Author: > As I mentioned above, this has the tendency to divorce errors from the context of the spans they were created in.

Huh, you're right! For some reason, my mental model has been that an anyhow::Error captures the span that it was created in, but it doesn't.

There is https://docs.rs/tracing-error/latest/tracing_error/ for capturing the current span, but anyhow won't use it automatically.

> you *must* log it (if it needs to be logged at all).

Reviewer 2: Raced to comment on `SpanTrace` in topmost thread: https://github.com/neondatabase/neon/pull/2681#issuecomment-1288554825 but this thread seems better suited.

After some [experimentation](https://gist.github.com/koivunej/6f6ac896ddea7f6be01245bb2ca56097) (includes example output in comment), `color-eyre` is required to format and `tracing-error` are required to capture the `SpanTrace`. The SPANTRACE header comes from [`color-spantrace`](https://github.com/yaahc/color-spantrace/blob/1a52aac2b55a05efc64807f06b91414550805af8/src/lib.rs#L353).

> you *must* log it (if it needs to be logged at all).

Reviewer 3: That's very cool! Seeing that, I wonder if it's possible to capture the `Span` object itself, so that logging errors can temporarily re-enter all of the necessary spans instead. If not, I think `SpanTrace` is still a great solution to this :)

For example:

```rust
fn read_motd_file() -> std::io::Result<String> {
    let mut f = File::open("/etc/motd")?;
    let mut result = String::new();
    f.read_to_string(&mut result)?;
    result
}
```

Opening or reading the file could fail, but there is no need to log
the error here. The function merely propagates the error to the
caller, and it is up to the caller to log the error or propagate it
further, if the failure is not expected. But if, for example, it is
normal that the "/etc/motd" file doesn't exist, the caller can choose

> to silently ignore the error, or log it in INFO or DEBUG level:

Reviewer: ```suggestion
to silently ignore the error, or log it as an INFO or DEBUG level message:
```

```rust
fn get_message_of_the_day() -> String {
    // Get the motd from /etc/motd, or return the default proverb
    match read_motd_file() {
        Ok(motd) => motd,
        Err(err)  => {
            // It's normal that /etc/motd doesn't exist, but if we fail to
            // read it for some other reason, that's unexpected. The message
            // of the day isn't very important though, so we just WARN and
            // continue with the default in any case.
            if err.kind() != std::io::ErrorKind::NotFound {

>                  tracing::warn!("could not read \"/etc/motd\": {err:?}");

Reviewer 4: Instead of just formatting the message, it is helpful for generating structured logs when relevant bits of information are recorded to separate fields. This particular example might not be the prime one for that, but here we could do:
```rust
warn!(file = "/etc/motd", "could not read {file}: {err:?}");
```

So the log output becomes(plain format):
```console
2022-10-23T15:09:32.007339Z  WARN could not read /etc/motd: PermissionDenied file=/etc/motd
```

and with json formatting we would have:
```json
{"timestamp":"2022-10-23T15:09:32.007339Z","level":"WARN","fields":{"message":"could not read /etc/motd: PermissionDenied","file":"/etc/motd"}}
```

We could distill a couple sentences about structuring and then have our examples follow that.

Reviewer 2: Re: my other https://github.com/neondatabase/neon/pull/2681#issuecomment-1288554825, I'd go as far as not including any variable context in the message:

```rust
warn!(?file, ?err, "read failed");
```

But will need to do some experimentation on what this would look like.

Reviewer 4: I would like to agree and partially do, but wouldn't suggest completely against having some context in the message either, because it's a bit less tool agnostic in sense `message` field becomes a bit less useful and then there needs to be a view for each type of logs. So would write the message for humans and separate the fields for computers, so that they can make things more accessible for humans.

            }
            "An old error is always more popular than a new truth. - German proverb"
        }
    }
}
```

## Error types

> We use the `anyhow` crate widely. It contains many convenient macros

Reviewer 5: I have a different interpretation which involves two layers of thinking:

* On the surface, `.context(..)` [is](https://docs.rs/anyhow/latest/src/anyhow/context.rs.html#90-105) `ok_or_else(|| anyhow!(...))`. Which means that every time I see the latter, the question pops up: "why this again?". The name _might_ (but read below) be controversial, but the method itself is convenient.

* > it's unclear when reading Option::context that it's the source of the error instead of something providing context to an existing error

Oh, but there's an error and it's `None`! This is the source we've been looking for (although it's very shy and prefers to keep its secrets). The method does what it claims to do: it attaches a context to _an error_. Even the `eyre`'s doc explicitly says that: _anyhow avoids this issue by using a **different mental model** where you’re adding “context” to an error, though this not a mental model for error handling that eyre agrees with._

Clearly, you're accustomed to the `eyre`'s model, but this doesn't necessarily mean that the other model is wrong. Especially considering the first bullet.

> We use the `anyhow` crate widely. It contains many convenient macros

Reviewer 3: > Clearly, you're accustomed to the `eyre`'s model, but this doesn't necessarily mean that the other model is wrong.

Fair enough :)

> Oh, but there's an error and it's `None`!

I think this is the part that I disagree with, but if I'm in the minority here that's ok

like `bail!` and `ensure!` to construct and return errors, and to
propagate many kinds of low-level errors, wrapped in `anyhow::Error`.

A downside of `anyhow::Error` is that the caller cannot distinguish
between different error cases. Most errors are propagated all the way
to the mgmt API handler function, or the main loop that handles a
connection with the compute node, and they are all handled the same

> way: the error is logged and returned the error to the client as

Reviewer: ```suggestion
way: the error is logged and returned to the client as
```

an HTTP or libpq error.

But in some cases, we need to distinguish between errors and handle
them differently. For example, attaching a tenant to the pageserver
could fail either because the tenant has already been attached, or
because we could not load its metadata from cloud storage. The first
case is more or less expected. The console sends the Attach request to
the pageserver, and the pageserver completes the operation, but the
network connection might be lost before the console receives the
response. The console will retry the operation in that case, but the
tenant has already been attached. It is important that the pagserver
responds with the HTTP 403 Already Exists error in that case, rather
than a generic HTTP 500 Internal Server Error.

If you need to distinguish between different kinds of errors, create a

> new `Error` type. The `thiserror` create is useful for that. But in

Reviewer: ```suggestion
new `Error` type. The `thiserror` crate is useful for that. But in
```

> most cases `anyhow::Error` is good enough.

Reviewer 6: I practice poor man's typing with anyhow 

https://github.com/neondatabase/neon/blob/9f49605041cc4954eff96690c7874cc16ac3f8f8/libs/remote_storage/src/lib.rs#L154-L162

which can indicate various failure scenarios with the enum variants.

> most cases `anyhow::Error` is good enough.

Reviewer 7: It seems that I'm in minority here. But I'll raise my hand anyway :)

To me error handling with anyhow is broken for anything other than simple CLIs which just bubble up errors to the user without trying to inspect them. 

The ideal approach to me is that every module defines its error enum with `thiserror::Error` (it just saves boilerplate on writing impls manually). It has the benefit that anyone who sees this module for the first time can see what are the possible errors. Listing error variants requires to put more thought into it compared to just a bunch of `bail!("A is broken")` calls. Better names for errors lead to better answers to "How I can handle this error?" type of questions. 

Nesting anyhow::Error types in other enum has following downsides:
* When there are many anyhow::Error variants `From` doesnt work anymore which leads to boilerplate since `?` is not available.
* If you need to detalize some anyhow::Error into new variants then you have to change every caller of a function returning such an error. With thiserror you need only add new variant and change the places which actually handle the error

This approach is more radical and requires ground up design for error types which is not an easy task

> most cases `anyhow::Error` is good enough.

Reviewer 8: See @funbringer 's comment https://github.com/neondatabase/neon/pull/2681/files#r1004581973

## Panics

Depending on where a panic happens, it can cause the whole pageserver
or safekeeper to restart, or just a single tenant. In either case,
that is pretty bad and causes an outage. Avoid panics. Never use
`unwrap()` or other calls that might panic, to verify inputs from the
network or from disk.

It is acceptable to use functions that might panic, like `unwrap()`, if
it is obvious that it cannot panic. For example, if you have just

> checked that a variable is not None, it is OK to call `unwrap()` on it.

Reviewer 6: ```suggestion
checked that a variable is not None, it is OK to call `unwrap()` on it, but still prefer to use `expect("reason")` instead to explain why it cannot fail.
```

Reviewer: "but it is still preferable  to use `expect("reason")` instead to explain why the function cannot fail."

Author: thanks, added!

`assert!` and `panic!` are reserved for checking clear invariants and
very obvious "can't happen" cases. When in doubt, use anyhow `ensure!`
or `bail!` instead.

## Error levels

`tracing::Level` doesn't provide very clear guidelines on what the
different levels mean, or when to use which level. Here is how we use
them:

### Error

Examples:
- could not open file "foobar"
- invalid tenant id

Errors are not expected to happen during normal operation. Incorrect
inputs from client can cause ERRORs. For example, if a client tries to
call a mgmt API that doesn't exist, or if a compute node sends passes
an LSN that has already been garbage collected away.

These should *not* happen during normal operations. "Normal
operations" is not a very precise concept. But for example, disk

> errors is not expected to happen when the system is working, so that

Reviewer: ```suggestion
errors are not expected to happen when the system is working, so those
```

> counts as an Error. However, if a TCP connection to a compute node is

Reviewer: ```suggestion
count as Errors. However, if a TCP connection to a compute node is
```

lost, that is not considered an Error, because it doesn't affect the
pageserver's or safekeeper's operation in any way, and happens fairly
frequently when compute nodes are shut down, or are killed abruptly
because of errors in the compute.

**Errors are monitored, and always need human investigation to determine

> the cause.**

Reviewer 3: Is this currently the case? I'd assume because of #2653 that it's not

Author: They are not monitored currently. This sentence is aspirational :-).

Whether something should be logged at ERROR, WARNING or INFO level can
depend on the callers and clients. For example, it might be unexpected

> and a sign of a serious issue, if the console calls the

Reviewer: ```suggestion
and a sign of a serious issue if the console calls the
```

"timeline_detail" mgmt API for a timeline that doesn't exist. ERROR
would be appropriate in that case. But if the console routinely calls
the API after deleting a timeline, to check if the deletion has

> completed, then it would be totally normal and INFO or DEBUG level

Reviewer: ```suggestion
completed, then it would be totally normal and an INFO or DEBUG level message
```

would be more appropriate. If a message is logged as an ERROR, but it
in fact happens frequently in production and never requires any
action, it should probably be demoted to INFO level.

### Warn

Examples:
- could not remove temporary file "foobar.temp"
- unrecognized file "foobar" in timeline directory

Warnings are similar to Errors, in that they should not happen
when the system is operating normally. The difference between Error and
Warning is that an Error means that the operation failed, whereas Warning

> means that something unexpected happened, but the operation continued anyway.

Reviewer 3: I really like this definition. It aligns with what I've intuitively been doing for a while.

For example, if deleting a file fails because the file already didn't exist,
it should be logged as Warning.

> **Note:** The python regression tests, under `test_regress`, check the
> pageserver log after each test for any ERROR and WARN lines. If there are
> any ERRORs or WARNs that have not been explicitly listed in the test as
> allowed, the test is marked a failed. This is to catch unexpected errors
> e.g. in background operations, that don't cause immediate misbehaviour in

> > the tested functionality.

Author: This is not in 'main' yet. It is being done in https://github.com/neondatabase/neon/pull/2653.

### Info

Info level is used to log useful information when the system is
operating normally. Info level is appropriate e.g. for logging state
changes, background operations, and network connections.

Examples:
- "system is shutting down"
- "tenant was created"
- "retrying S3 upload"

### Debug & Trace

> Debug and Trace level messages are not Debug-level messages are not

Reviewer 2: ```suggestion
Debug and Trace level messages are not
```

printed to the log in our normal production configuration, but could
be enabled for a specific server or tenant, to aid debugging.

> (Although we don't actually have that capability as of this writing).

Reviewer 3: It'd be good to have more information on what the distinction between Debug and Trace should be, and when to use them instead of Info

Reviewer 4: There was a brief discussion whether we should adopt https://github.com/trentm/node-bunyan#levels since there would have been a rust library following the same scheme. Although didn't necessarily warm up to idea to introduce one more level, because that would mean going through pretty much any existing error to figure out if it's error or fatal, the bunyan definitions could have some useful points.

Author: I like those node-bunyan definitions, and an additional FATAL level also makes sense to me. However, `tracing` doesn't have FATAL, so I don't think we can adopt that.

## Context

We use logging "spans" to hold context information about the current
operation. Almost every operation happens on a particular tenant and
timeline, so we enter a span with the "tenant_id" and "timeline_id"
very early when processing an incoming API request, for example. All

> background operations should also run in a span contain at least those

Reviewer: ```suggestion
background operations should also run in a span containing at least those
```

two fields, and any other parameters or information that might be
useful when debugging an error that might happen when performing the
operation.

## Error message style

PostgreSQL has a style guide for writing error messages:

https://www.postgresql.org/docs/current/error-style-guide.html

Follow that guide when writing error messages in the PostgreSQL
extension. We don't follow it strictly in the pageserver and
safekeeper, but the advice in the PostgreSQL style guide is generally

> good, and you can't go wrong by following it.

Reviewer: +1
Microsoft also provides decent error message writing guidelines and best practices. https://learn.microsoft.com/en-us/windows/win32/debug/error-message-guidelines
We might also give some thought to what an error message review process would look like. I have found such reviews to occur upon request, periodically, or before a feature is released. I am not sure what approach makes the most sense for Neon, but I would be interested in any thoughts on this.

Author: > Microsoft also provides decent error message writing guidelines and best practices. https://learn.microsoft.com/en-us/windows/win32/debug/error-message-guidelines

Good advice

> We might also give some thought to what an error message review process would look like. I have found such reviews to occur upon request, periodically, or before a feature is released. I am not sure what approach makes the most sense for Neon, but I would be interested in any thoughts on this.

For the storage components, I think any PR that includes error messages is the right place to review the messages too. A separate review would draw more attention to error messages and make them better, for sure, but I don't think it's worth the effort. (For the user-facing web UI and APIs, error messages are much more prominent and more attention might be warranted.)
